src/login/logind-session.c

```diff
@@ -697,6 +699,71 @@ static int session_start_scope(Session *s, sd_bus_message *properties, sd_bus_er
         return 0;
 }
 
+static int session_dispatch_stop_on_idle(sd_event_source *source, uint64_t t, void *userdata) {
+        Session *s = userdata;
+        dual_timestamp ts;
+        usec_t now;
+        int r, idle;
+
+        assert(s);
+
+        if (s->stopping)
+                return 0;
+
+        r = sd_event_now(s->manager->event, CLOCK_MONOTONIC, &now);
+        if (r < 0)
+                return r;
+
+        idle = session_get_idle_hint(s, &ts);
+        if (idle) {
+                log_debug("Session \"%s\" of user \"%s\" is idle, stopping.", s->id, s->user->user_record->user_name);
+
+                r = session_stop(s, /* force */ true);
+                if (r < 0)
+                        log_error_errno(r, "Failed to stop idle session \"%s\": %m", s->id);
+                return r;
+        }
+
+        r = sd_event_source_set_time(source,  ts.monotonic + s->manager->stop_idle_session_usec);
```

Comment:
I will fix double white-space. However, I will keep `sd_event_source_set_time()` because we want to schedule next wake up not in relation to the current time but to the idle hint timestamp of this session.